src/core/bpf-foreign.c

```diff
@@ -0,0 +1,213 @@
+
+        *ret_attach_type = (enum bpf_attach_type) attach_type;
+        *ret_bpffs_path = TAKE_PTR(dup);
+
+        return 0;
+}
+
+int bpf_foreign_program_to_string(enum bpf_attach_type attach_type, const char *bpffs_path, char **ret_str) {
+        _cleanup_free_ char *p = NULL, *x = NULL;
+        const char *s = NULL;
+
+        assert(bpffs_path);
+        assert(ret_str);
+
+        s = bpf_cgroup_attach_type_to_string(attach_type);
+
+        /* Likely should update bpf_cgroup_attach_type_table with new mapping. */
+        if (!s) {
+                if (asprintf(&x, "unknown(%d)", attach_type) < 0)
+                        return log_oom();
+                s = x;
+        }
+
+        p = strjoin(s, ":", bpffs_path);
+        if (!p)
+                return log_oom();
+
+        *ret_str = TAKE_PTR(p);
+        return 0;
+}
+
+int bpf_foreign_prepare(Unit *u, enum bpf_attach_type attach_type, const char *bpffs_path) {
+        _cleanup_(bpf_program_unrefp) BPFProgram *prog = NULL;
+        _cleanup_free_ BPFForeignKey *key = NULL;
+        uint32_t prog_id;
+        int r;
+
+        assert(u);
+        assert(bpffs_path);
+
+        r = bpf_program_new_from_bpffs_path(bpffs_path, &prog);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to create foreign BPFProgram: %m");
+
+        r = bpf_program_get_id_by_fd(prog->kernel_fd, &prog_id);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to get BPF program id by fd: %m");
+
+        r = bpf_foreign_key_new(prog_id, attach_type, &key);
+        if (r < 0)
+                return log_unit_error_errno(u, r,
+                                "Failed to create foreign BPF program key from path '%s': %m", bpffs_path);
+
+        if (hashmap_contains(u->bpf_foreign_next_by_key, key)) {
+                log_unit_warning_errno(u, SYNTHETIC_ERRNO(EEXIST),
+                                "Foreign BPF program already exists, ignoring: %m");
+                return 0;
+        }
+
+        r = hashmap_ensure_put(&u->bpf_foreign_next_by_key, &bpf_foreign_next_by_key_hash_ops, key, prog);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to put foreign BPFProgram into map: %m");
+
+        TAKE_PTR(key);
+        TAKE_PTR(prog);
+
+        return 0;
+}
+
+int bpf_foreign_install(Unit *u) {
+        _cleanup_free_ char *cgroup_path = NULL;
+        int r;
+
+        assert(u);
+
+        if (!unit_get_cgroup_context(u))
+                return 0;
+
+        r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, NULL, &cgroup_path);
+        if (r < 0)
+                return r;
+
+        set_clear(u->bpf_foreign_installed);
+        r = attach_programs(u, cgroup_path, u->bpf_foreign_next_by_key, &u->bpf_foreign_installed, BPF_F_ALLOW_MULTI);
+        if (r < 0)
+                  log_unit_error_errno(u, r, "Failed to install foreign BPF programs: %m");
+
+        bpf_foreign_reset(u);
+
+        return r;
```

Comment:
this function logs in some cases at error level but not in others. as per coding style we want "library-like" functions to not log at all beyond debug typically. In general: either a function should log about all errors above debug level or about none (in which case it's the caller's responsibility to log if nedeed).

We require this so that we don't accidentally log twice about the same error (which is confusing at the least) or none at all (which is hard to debug then)

Hence, please figure out if your functions are more "library-like" or more "main-program-like". In the first case downgrade the log messages to debug or remove them and let the caller log. in the latter case, make sure you log in all error paths so that the caller doesn't log anymore.

usually stuff in src/basic/ and src/shared/ is — with exceptions — mostly library-like, and stuff in src/core/ can be either, depending if it's more "support" code or more "main" code. See https://systemd.io/CODING_STYLE/#error-handling

Reply:
Switched `src/core` functions to "main-program-like" style.